### DISCUSSION TITLE
Skipped job = failed job to failure()?

### DISCUSSION THREAD

**Discussion by altjx** [community/community#25769] [Actions]
<p>I have two jobs that is supposed to only run if any previous job fails, or if there’s a label attached to the PR. However, this job seems to <em>always</em> run, even if no label exists and if the previous jobs don’t fail (but are skipped instead).</p>
<p>Here are the two jobs:</p>
<pre><code class="lang-auto">  remove_staging_label_on_fail:
    needs: [build_app, build_sidekiq]
    if: ${{ failure() }} &amp;&amp; contains(github.event.pull_request.labels.*.name, 'ready for staging')
    runs-on: self-hosted
    steps:
      - uses: actions-ecosystem/action-remove-labels@v1
        with:
          labels: |
            ready for staging

  remove_prod_label_on_fail:
    needs: [build_app, build_sidekiq]
    if: ${{ failure() }} &amp;&amp; contains(github.event.pull_request.labels.*.name, 'ready for production')
    runs-on: self-hosted
    steps:
      - uses: actions-ecosystem/action-remove-labels@v1
        with:
          labels: |
            ready for production
</code></pre>
<p>Obviously they’re very close except the label that gets removed depends on the label that was previously attached.</p>
<p>Here’s a summary output of my workflow:</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/3X/7/d/7de9c65f8c7d96a6c410162b2641d7b44f09a2cd.jpeg" data-download-href="/uploads/short-url/hXSxJWpvJjgCXiAuJafIV3dAeFn.jpeg?dl=1" title="Screen Shot 2021-12-29 at 10.00.25 PM" rel="noopener nofollow ugc"><img src="https://user-images.githubusercontent.com/4023423/181093183-f9748767-3844-4b8c-ab2c-4b970421d381.jpeg" alt="Screen Shot 2021-12-29 at 10.00.25 PM" data-base62-sha1="hXSxJWpvJjgCXiAuJafIV3dAeFn" width="690" height="340" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/7/d/7de9c65f8c7d96a6c410162b2641d7b44f09a2cd_2_690x340.jpeg, https://aws1.discourse-cdn.com/github/optimized/3X/7/d/7de9c65f8c7d96a6c410162b2641d7b44f09a2cd_2_1035x510.jpeg 1.5x, https://aws1.discourse-cdn.com/github/original/3X/7/d/7de9c65f8c7d96a6c410162b2641d7b44f09a2cd.jpeg 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/7/d/7de9c65f8c7d96a6c410162b2641d7b44f09a2cd_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use href="#far-image"></use></svg><span class="filename">Screen Shot 2021-12-29 at 10.00.25 PM</span><span class="informations">1376×680 74.9 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use href="#discourse-expand"></use></svg>
</div></a></div>
<p>As you can see above, nothing failed. Also again, there is no label in my PR. I’m very curious to know why these two jobs ran given this scenario.</p>

**Comment by airtower-luna** [comment, marked as answer]
<aside class="quote no-group" data-username="altjx" data-post="1" data-topic="220273">
<div class="title">
<div class="quote-controls"></div>
 altjx:</div>
<blockquote>
<p><code>    if: ${{ failure() }} &amp;&amp; contains(github.event.pull_request.labels.*.name, 'ready for production')</code></p>
</blockquote>
</aside>
<p>It’s because you enclosed only <code>failure()</code> in <code>${{ ... }}</code>, not the whole expression (or nothing). That way everything outside the <code>${{ ... }}</code> is treated as a string, and a non-empty string evaluates to <code>true</code>.</p>
<p>Fix:</p>
<pre><code class="lang-yaml">    if: ${{ failure() &amp;&amp; contains(github.event.pull_request.labels.*.name, 'ready for production') }}
</code></pre>

**Comment by altjx** [comment]
<p>Ahhh I gotcha!! Thanks so much.</p>
<p>Just out of curiosity, what exactly is <em>true</em> in that condition that actually triggers it, based on the way it is now?</p>

**Comment by airtower-luna** [comment]
<p>Depending on the value returned by <code>failure()</code> either</p>
<pre><code class="lang-auto">"false &amp;&amp; contains(github.event.pull_request.labels.*.name, 'ready for production')"
</code></pre>
<p>or</p>
<pre><code class="lang-auto">"true &amp;&amp; contains(github.event.pull_request.labels.*.name, 'ready for production')"
</code></pre>
<p>and it’s a string either way. :wink:</p>

**Comment by altjx** [comment]
<p>Ahh ok got it. And it seems that <code>if "string"</code> returns <code>true</code> so that clarifies what’s going on lol.</p>
<p>Thanks a lot!!</p>
